util/jsonrpc/jsonrpc.go

```diff
@@ -294,37 +289,63 @@ func (client *RPCClient) setConnectionError(err error) {
 // desktop applications, but we store the active connection and ping it regularly to
 // keep it alive (see ping()).
 func (client *RPCClient) conn() (*connection, error) {
-	if client.connection == nil {
-		defer client.connLock.Lock()()
-		if client.connection == nil {
-			defer client.backendsLock.RLock()()
-			start := 0
-			if len(client.backends) > 0 {
-				start = rand.Intn(len(client.backends))
-			}
-			var lastErr error
-			for i := 0; i < len(client.backends); i++ {
-				client.log.Debugf("Trying to connect to backend %v", client.backends[start].Name)
-				lastErr = client.establishConnection(client.backends[start])
-				if lastErr != nil {
-					client.log.WithError(lastErr).Info("Failover: backend is down")
-					start = (start + 1) % len(client.backends)
-				} else {
-					client.log.Debug("Successfully connected to backend")
-					break
-				}
-			}
-			if lastErr != nil {
-				client.log = client.log.WithField("backend", "offline")
-				// tried all backends
-				client.log.Debug("Disconnected from all backends")
-				client.setConnectionError(lastErr)
-				return nil, lastErr
+	client.connLock.RLock()
+	conn := client.connection
+	client.connLock.RUnlock()
+	if conn != nil {
+		return conn, nil
+	}
+
+	conn, err := func() (*connection, error) {
+		client.connLock.Lock()
+		defer client.connLock.Unlock()
+		if client.connection != nil {
+			return client.connection, nil
+		}
+
+		start := 0
+		if len(client.backends) > 0 {
+			start = rand.Intn(len(client.backends))
+		}
+		var lastErr error
+		for i := 0; i < len(client.backends); i++ {
+			client.log.Debugf("Trying to connect to backend %v", client.backends[start].Name)
+			conn, err := client.establishConnection(client.backends[start])
+			if err != nil {
+				client.log.WithError(err).Info("Failover: backend is down")
+				start = (start + 1) % len(client.backends)
+				lastErr = err
+				client.connection = nil
+			} else {
+				client.log.Debug("Successfully connected to backend")
+				client.connection = conn
+				lastErr = nil
+				break
 			}
-			client.setConnectionError(nil)
 		}
+		if lastErr != nil {
+			client.log = client.log.WithField("backend", "offline")
+			// tried all backends
+			client.log.Debug("Disconnected from all backends")
+			client.setConnectionError(lastErr)
+			return nil, lastErr
+		}
+		client.setConnectionError(nil)
+		return client.connection, nil
+	}()
```

Comment:
`client.connLock` is released here but `client.connection` isn't ready yet. I think this is still a source of bugs.

Ideally, you'd hold `client.connLock` for the whole duration of the method but you can't because `client.onConnectCallback()` runs the electrum protocol negotiation which needs `client.connection`. This is the part which I think doesn't belong to the RPC client. Some time later in a rewrite.

Reply:
Yep I noticed this. If the negotiation fails, the connection is also not closed and the read loop runs forever...

Some day! :pray: